11-graphs/States/main.c

```diff
@@ -0,0 +1,98 @@
+#define _CRT_SECURE_NO_WARNINGS
+
+#include "graphsTests.h"
+#include "graph.h"
+#include "testsForQueue.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+
+Graph* buildGraph(const char* fileName, bool* errorCode) {
+    FILE* file = fopen(fileName, "r");
+    int numberOfVertices = 0;
+    fscanf(file, "%d ", &numberOfVertices);
+    Graph* graph = createGraph(numberOfVertices, errorCode);
+    if (*errorCode) {
+        return NULL;
+    }
+
+    int numberOfEdges = 0;
+    fscanf(file, "%d ", &numberOfEdges);
+    int numberOfVerticesCreated = 0;
+    for (int i = 0; i < numberOfEdges; ++i) {
+        VertexValue value = { .isCapital = false, .key = -1, .stateNumber = -1 };
+        int firstKey = -1;
+        fscanf(file, "%d ", &firstKey);
+        value.key = firstKey;
+        addVertex(graph, value, errorCode);
+        if (*errorCode) {
+            deleteGraph(&graph, errorCode);
+            return NULL;
+        }
+
+        int secondKey = -1;
+        fscanf(file, "%d ", &secondKey);
+        value.key = secondKey;
+        addVertex(graph, value, errorCode);
+        if (*errorCode) {
+            deleteGraph(&graph, errorCode);
+            return NULL;
+        }
+
+        int edgeWeight = -1;
+        fscanf(file, "%d ", &edgeWeight);
+
+        graph = connectVertices(graph, firstKey, secondKey, edgeWeight, errorCode);
+        if (*errorCode) {
+            deleteGraph(&graph, errorCode);
+            return NULL;
+        }
+    }
+
+    int numberOfCapitals = -1;
+    fscanf(file, "%d ", &numberOfCapitals);
+    for (int i = 0; i < numberOfCapitals; ++i) {
+        int city = -1;
+        fscanf(file, "%d ", &city);
+        setCapital(graph, city, errorCode);
+    }
+    return graph;
+}
+
+int main(void) {
+    bool errorCode = false;
+    testGraph(&errorCode);
+    if (errorCode) {
+        return errorCode;
+    }
+    testQueue(&errorCode);
+    if (errorCode) {
+        return errorCode;
+    }
```

Comment:
```suggestion
    testGraph(&errorCode);
    testQueue(&errorCode);
    if (errorCode) {
        return errorCode;
    }
```
Тесты ведь независимы друг от друга, если хоть один не пройдёт, то всё, но почему бы им обоим не дать поработать

Reply:
Но если мы хотим, чтобы очередь тоже тестировалась, разве не нужно сбросить код ошибки? И как понять тогда прошли ли тесты для графов? Получается нужно результаты тестов в отдельных переменных хранить? Кстати, графы использую очередь, можно сказать, что если тесты для очереди не проходят, то графы можно не проверять?

Reply:
А мне казалось, что у Вас намеренно нигде errorCode не выставляется в true, чтобы если он единожды стал false, он оставался false. И тесты на errorCode не смотрят, они просто выставляют его в false, если что-то пошло не так. Если это неправда, то, может, стоит так сделать.

В реальной жизни общая политика такая, что запускаем все тесты, какие можем, потом разбираемся (правда, в реальной жизни использовались бы mock-объекты, чтобы разорвать связь между графом и очередью). Это хотя бы даст шанс за один тестовый прогон выявить ошибки в графе (может, он падает, не дойдя до очередей). Почему — тестов обычно много и гоняются они долго, так что чтобы поправить ошибку и перезапустить, может потребоваться следующего дня ждать.

Reply:
errorCode выставляется в true в случае какой-либо ошибки. Я думал, если что-то пошло не так, нужно выставить true и завершить программу, почему тесты должны выставлять его в false, если что-то не так?

Reply:
Значит, в рассуждении выше наоборот, true -> false. Если произошла какая-то ошибка, errorCode оказывается в состоянии "ошибка" и никак не может из него выйти (что хорошо и правильно), поэтому мы смело можем запускать все остальные тесты, ошибка не потеряется.